### ISSUE TITLE
Wrong extension type argument for promoted/bounded extensions

### ISSUE THREAD

**Issue by johnniwinther** [dart-lang/sdk#43177]
Consider the following program currently accepted by analyzer and CFE:
```Dart
extension Extension<T> on T {
  T method1() => this;
}
extension BoundExtension<T extends Class> on T {
  T method2() => this;
}
class Class {}
class SubClass extends Class {}
Class test1<T>(T t1) {
  if (t1 is SubClass) {
    return t1.method1(); // 
  }
  return new Class();
}
test2<T extends Class>(T t2) {
  if (T == SubClass) {
    SubClass subClass = t2.method2();
  }
}
test3<T>(T t3) {
  if (t3 is SubClass) {
    SubClass subClass = t3.method2();
  }
}
```
* The static type of `t1` in `t1.method1()` is `T & SubClass`. Given that we cannot use the promoted type as a type argument to the extension method call, the argument should be `T`, i.e. we should transform `t1.method1()` into `Extension<T>(t1).method1()`. This means that the return type of the call should be `T` which is not assignable to `Class`. CFE (and analyzer, it seems) currently uses `SubClass` as type argument and therefore doesn't produce an error.
* The static type of `t2` in `t2.method2()` is `T` and therefore the type argument to the extension method call should be `T`, i.e. we should transform `t2.method2()` into `BoundExtension<T>(t2).method2()`. This means that the return type of the call should be `T` which is not assignable to `SubClass`. CFE (and analyzer, it seems) currently uses `Class` as type argument (and adds an implicit cast to `SubClass`) and therefore doesn't produce an error.
* The static type of `t3` in `t3.method3()` is `T & SubClass`. Given that we cannot use the promoted type as a type argument to the extension method call, the argument should be `T`, i.e. we should transform `t3.method2()` into `BoundExtension<T>(t3).method2()` which in turn should produce a compile-time error since `T` is not a valid argument for the bound on `BoundExtension`. CFE (and analyzer, it seems) currently uses `SubClass` as type argument and therefore doesn't produce an error.

**Comment by johnniwinther** [comment]
cc @eernstg @lrhn Do you agree with the analysis?

**Comment by eernstg** [comment]
I agree with the analysis. I think the underlying conflict is that the non-reifiable intersection type `X & U` is always erased to `X`, never to `U`, and it is in fact more convenient to have the type `U` in some situations.

**Comment by lrhn** [comment]
I agree with the analysis too.

If there is any way we can keep reifying `T&SubClass` into ``SubClass` when used in a context where a `SubClass` is required and `T` does not fit, I'd like to do that. We have all the information, and we can see that choosing `T` is going to be invalid, so let's move down the `&` chain until we find something which does work.
That would also meant that if you have a static type of `T&R&S&Foo` which is needed as a type in a context of `X extends Bar` , then we'll check whether `T` extends `Bar` (based on the bound), then `R`, then `S`, and finally `Foo`, and pick first one which satisfies the constraint. (If that isn't what we already do).

Obviously, and as usual, I can't rule out that doing that will make the type system Turing complete, in which case we probably don't want it anyway.

**Comment by johnniwinther** [comment]
It might be that for the third case we should never resolve `t3.method2()` to the extension because the inferred type argument (`T` instead `T & SubClass`) is not a valid argument for `BoundExtension`. At leasts that is what happens when I fix the case in the CFE.

**Comment by eernstg** [comment]
> we should never resolve `t3.method2()` to the extension

Right, the extension is not applicable in the first place, so it's an error about `method2` not resolving to anything.

**Comment by johnniwinther** [comment]
I have a CL ready for this for the CFE: https://dart-review.googlesource.com/c/sdk/+/160524

**Comment by scheglov** [comment]
> we cannot use the promoted type as a type argument to the extension method call

Could you remind me why?

Looking into the [type inference PR](https://github.com/dart-lang/language/pull/957/files#diff-8799cf34fe82cd1bb7ba3814a5e7fca4R740), it seems that we match using promoted bounds as well.

**Comment by lrhn** [comment]
The promoted type is an intersection type `T&SubClass`. The type argument to the extension can be used as a type inside the extension, and we do not allow intersection types to be reified as type variables or the declared types of variables. Intersection types can only occur as the type of promoted variables.

So, we have to throw away the intersection part of the type when reifying it as `Extension<T>(x).method1()`.

The (useful) mistake here was to use `SubClass` instead of `T` as that type.
